### PR TITLE
[CSSolver] Use correct locator when matching function result types re…

### PR DESCRIPTION
**Pull request by xedin**
…lated to closures

Currently we always use 'FunctionResult' as a path element when matching
function result types, but closure result type is allowed to be implicitly
converted to Void, which means we need to be careful when to use
'FunctionResult' and 'ClosureResult'.

Resolves: rdar://problem/37790062

### CODE REVIEW

**Comment by rudkx** [comment]
fyi, i'm taking a careful look at this

**Comment by rudkx** [comment]
I think this seems reasonable if slightly precarious. I suspect I am going to end up substantially reworking/replacing `getPotentialBindings()` in the not-too-distant-future and hopefully as a result of that make something that doesn't have all the strange ordering issues we have today.

**Comment by xedin** [comment]
@rudkx Ok, sounds good! I'll try to figure out what do we need to do to make this work in `CSApply` then.

**Comment by xedin** [comment]
@rjmccall @jckarter  Looks like I'd need some help from you here, it seems like `ResultPlan` doesn't account for conversions from `() -> T` to `() -> Void` although we create `FunctionConversionExpr` to mark that... 

I can fix up the assert at https://github.com/apple/swift/blob/master/lib/SILGen/SILGenPoly.cpp#L1880, but when I do so it fails at
https://github.com/apple/swift/blob/master/lib/SILGen/SILGenPoly.cpp#L1940 to cast `innerSubstType` to tuple (which `innerSubstType` is not, it's an archetype related to `A`) but it tries to emit it into empty tuple (which is what `Void` type is).

So I'm wondering what would be the best approach here? I guess since it's trying to convert function which returns something to function which returns nothing it could be fine to just drop the result type on the floor?

**Comment by xedin** [comment]
Simpler example:

```swift
struct S<T> {
  init(_ a: () -> T, _ b: () -> T) {}
}

func foo() -> Int { return 42 }
func bar() -> Void {}

_ = S({ foo() }, { bar() })
```

**Comment by jckarter** [comment]
Dropping a result isn't always an ABI-compatible function conversion if the result is returned indirectly or needs to be released. We ought to be emitting a thunk when this conversion happens. The failures you're seeing suggest to me that isn't happening.

**Comment by xedin** [comment]
@jckarter Sorry if it wasn’t clear but it fails while trying to emit such a thunk, it seems like it doesn’t support converting result from something to void...

**Comment by xedin** [comment]
@jckarter What is the best way to express in result planner conversion from something to void?

**Comment by jckarter** [comment]
If we're converting a result to void, then we need to destroy the result value(s) and produce an empty tuple.

**Comment by xedin** [comment]
@jckarter Could you please elaborate where I could start to figure out how to do that? :)

**Comment by xedin** [comment]
@swift-ci please smoke test

**Comment by xedin** [comment]
@swift-ci please smoke test

File: lib/SILGen/SILGenPoly.cpp [hunk 1]
```diff
@@ -1876,6 +1883,29 @@ void ResultPlanner::plan(AbstractionPattern innerOrigType,
...
+          SGF.silConv.isSILIndirect(innerResult))
+        (void)addInnerIndirectResultTemporary(planData, innerResult);
+      else
+        addIgnoreDirectResult();
+    };
+
+    if (innerOrigType.isTuple()) {
+      auto innerTuple = cast<TupleType>(innerSubstType);
+      for (unsigned i = 0, n = innerTuple->getNumElements(); i != n; ++i)
+        ignoreNextResult();
```
Comment (rjmccall) [inline lib/SILGen/SILGenPoly.cpp]: 1. This needs be recursive — we expand tuples recursively, so this needs to do it, too.
2. This needs to follow the inner abstraction pattern, not the inner substituted type.  There could be a generic type that's been substituted to a tuple.

I would suggest breaking this out as a method on the planner.

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test source compatibility

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test source compatibility

**Review by rjmccall — APPROVED** [review]
Thanks, that looks good.

**Pull request merged**